Samza runner support for non unique stateId across multiple ParDos

seungjin-an opened the pull request:
Samza runner support for non-unique stateIds across multiple ParDos
(1) Do not touch existing RocksDB Store Id with stateId used only by one ParDo
(2) Add a prescan logic in SamzaRunner to traverse Beam pipeline topologically and identify non-unique stateIds. 
(3) Using the set of non-unique stateIds, introduce a new mapping of user-provided stateIds to store ID 

// stateId used in a single ParDo continue to use the same stateId as storeId
.apply("First stateful ParDo", ParDo.of(fn)); 
→ “foo"

// stateId used in multiple ParDos use a new mapping to store Id with PTransform name
.apply("Second stateful ParDo", ParDo.of(fn2)); 
→ “foo-Second_stateful_ParDo”

(4) Re-enable ParDoTest in :runners:samza build.gradle

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

dxichen commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:
```diff
@@ -35,12 +34,17 @@ public class ConfigContext {
   private final SamzaPipelineOptions options;
-  private final Set<String> stateIds;
+  private final Map<String, String> usedStateIdMap;
+  private final Map<String, String> multiParDoStateIdMap;
```
Lets rename this for clarity stateIdsToRewrite the value could also be the new state id.

seungjin-an replied:
Sounds good - made the change.

dxichen commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java:
```diff
@@ -249,6 +252,19 @@ public String getTransformId() {
+            storeIds.put(
+                stateId,
+                multiParDoStateIds.contains(stateId)
+                    ? String.join("-", stateId, escapedParDoName)
```
Let's create a helper method for this join, looks like it happens in a lot of place in the code.

seungjin-an replied:
Done. Thanks!

dxichen reviewed (changes requested):
Minor style/readability comments, thanks for the diff!

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java:
```diff
@@ -140,9 +141,21 @@ public SamzaPipelineResult run(Pipeline pipeline) {
     LOG.info("Beam pipeline JSON graph:\n{}", jsonGraph);
 
     final Map<PValue, String> idMap = PViewToIdMapper.buildIdMap(pipeline);
+    /* Map of stateId to sanitized (remove whitespace and replace '-' with '_') PTransform name, used in multiple ParDos
```
Moved the javadoc with /**.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java:
```diff
@@ -105,6 +106,38 @@ public static void createConfig(
+          // put new config with multi pardo config
+          configBuilder.put(
+              "stores." + multiParDoStateId + ".factory",
+              "org.apache.samza.storage.kv.RocksDbKeyValueStorageEngineFactory");
```
Hi Daniel, could you elaborate more on what you mean by "already including the package?" Thanks! We do have some occurences of using these classnames in other places where we add the configs.

dxichen replied:
Oh I just mean something like `RocksDbKeyValueStorageEngineFactory.class.getName()`

seungjin-an replied:
Ah Sounds good :) Made the change to all other occurrences.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
         executableStage.getUserStates().stream()
             .map(UserStateReference::localName)
-            .collect(Collectors.toSet());
+            .collect(Collectors.toMap(Function.identity(), Function.identity()));
```
Thanks, made the change!

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -129,18 +130,7 @@ static KeyValueStore<ByteArray, StateValue<?>> getBeamStore(TaskContext context)
-    return createStateInternalsFactory(id, null, context, pipelineOptions, Collections.emptySet());
-  }
-
-  static <K> Factory<K> createStateInternalsFactory(
```
As we fetch stateIds from DoFnSignature earlier in the logic for handling stateIds with multiParDo, I've replaced this with using stateIdMap (map of stateId to RocksDB storeId). Would we prefer to add this back for future use?

dxichen replied:
got it, lets leave this out if it is no longer used

xinyuiscool reviewed (commented):
Two main issues:

1) no rewriting of configs, instead, do a prescan of pipeline and build the non-unique state id set
2) the changes in the SamzaStateInternals are not needed if you make the stores map still follow the same definition of {stateId -> store}. The state internals is generated per PTransform so it will not have dups here.

seungjin-an commented:
> Two main issues:
> 
> 1. no rewriting of configs, instead, do a prescan of pipeline and build the non-unique state id set
> 2. the changes in the SamzaStateInternals are not needed if you make the stores map still follow the same definition of {stateId -> store}. The state internals is generated per PTransform so it will not have dups here.

Thank you so much for taking the time reviewing, providing suggestions on the pre-scan/ better naming, and fixing my understanding in SamzaStoreInternals. I have addressed all the comments and please have another look.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:
```diff
@@ -35,12 +35,17 @@ public class ConfigContext {
+  private final Map<String, String> stateIdsToRewrite;
 
-  public ConfigContext(Map<PValue, String> idMap, SamzaPipelineOptions options) {
+  public ConfigContext(
```
Thank you for the suggestion here, made the change.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java:
```diff
@@ -140,9 +141,14 @@ public SamzaPipelineResult run(Pipeline pipeline) {
     LOG.info("Beam pipeline JSON graph:\n{}", jsonGraph);
 
     final Map<PValue, String> idMap = PViewToIdMapper.buildIdMap(pipeline);
+    final Map<String, String> multiParDoStateIdMap = new HashMap<>();
```
I totally agree - this approach seems much better. Made the change, thank you!

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java:
```diff
@@ -148,7 +149,8 @@ public DoFnOp(
       DoFnSchemaInformation doFnSchemaInformation,
-      Map<?, PCollectionView<?>> sideInputMapping) {
+      Map<?, PCollectionView<?>> sideInputMapping,
+      Map<String, String> userStateIds) {
```
As this is no longer needed, I reverted this diff but definitely agree. Thank you

seungjin-an replied:
Made the change.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaDoFnRunners.java:
```diff
@@ -92,14 +90,14 @@ public static <InT, FnOutT> DoFnRunner<InT, FnOutT> create(
       OpEmitter emitter,
-      FutureCollector futureCollector) {
+      FutureCollector futureCollector,
+      Map<String, String> userStateIds) {
```
Sounds good

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -99,13 +96,16 @@
   // the stores include both beamStore for system states as well as stores for user state
   private final Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores;
+  // Map of non-system stateIds to unique storeId
+  private final Map<String, String> stateIdMap;
```
Sounds good.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
-          stateId ->
-              stores.put(
-                  stateId, (KeyValueStore<ByteArray, StateValue<?>>) context.getStore(stateId)));
+      stateIdMap
```
Got it - thank you.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -99,13 +96,16 @@
 
   private SamzaStoreStateInternals(
       Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores,
+      Map<String, String> stateIds,
```
Ack removed the changes in this class

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -99,13 +96,16 @@
   private final String stageId;
 
   private SamzaStoreStateInternals(
       Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores,
```
Ack removed the changes in this class

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
+          .values()
+          .forEach(
+              storeId ->
+                  stores.put(
```
Ack removed the changes in this class

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
       stateKeyCoder = (Coder<K>) VoidCoder.of();
     }
-    return new Factory<>(Objects.toString(id), stores, stateKeyCoder, batchGetSize);
+    return new Factory<>(Objects.toString(id), stores, stateIdMap, stateKeyCoder, batchGetSize);
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -271,16 +263,19 @@ private static ByteArrayOutputStream getThreadLocalBaos() {
   public static class Factory<K> implements StateInternalsFactory<K> {
     private final String stageId;
     private final Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores;
+    private final Map<String, String> stateIdMap;
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -271,16 +263,19 @@ private static ByteArrayOutputStream getThreadLocalBaos() {
     public Factory(
         String stageId,
         Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores,
+        Map<String, String> stateIdMap,
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -271,16 +263,19 @@ private static ByteArrayOutputStream getThreadLocalBaos() {
         int batchGetSize) {
       this.stageId = stageId;
       this.stores = stores;
+      this.stateIdMap = stateIdMap;
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -303,7 +298,8 @@ public StateInternals stateInternalsForKey(@Nullable K key) {
       }
 
-      return new SamzaStoreStateInternals<>(stores, key, baos.toByteArray(), stageId, batchGetSize);
+      return new SamzaStoreStateInternals<>(
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -323,11 +319,11 @@ protected AbstractSamzaState(
       this.namespace = namespace;
       this.addressId = address.getId();
-      this.isBeamStore = !stores.containsKey(address.getId());
+      this.isBeamStore = !stores.containsKey(stateIdMap.get(address.getId()));
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:
```diff
@@ -323,11 +319,11 @@ protected AbstractSamzaState(
           isBeamStore
               ? (KeyValueStore) stores.get(BEAM_STORE)
-              : (KeyValueStore) stores.get(address.getId());
+              : (KeyValueStore) stores.get(stateIdMap.get(address.getId()));
```
Done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:
```diff
@@ -83,6 +84,10 @@ public void putAll(Map<String, String> properties) {
     config.putAll(properties);
   }
 
+  public void remove(String name) {
```
Makes sense - done.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:
```diff
@@ -64,8 +69,39 @@ public SamzaPipelineOptions getPipelineOptions() {
+            StoreIdUtils.toMultiParDoStoreId(stateId, prevParDoName);
+        usedStateIdMap.put(prevMultiParDoStateId, prevParDoName);
+        // Store the stateId with previous parDo name which will be used for config rewriting
+        stateIdsToRewrite.put(stateId, prevParDoName);
```
Changed the logic to prescanning.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java:
```diff
@@ -377,25 +392,21 @@ public Map<String, String> createConfig(
+        final String userStateId = state.id();
+        final String escapedParDoName =
+            SamzaPipelineTranslatorUtils.escape(node.getEnclosingNode().getFullName());
+        final String uniqueStoreId = ctx.getUniqueStoreId(userStateId, escapedParDoName);
```
Only need config creation part for translation so removed this part.

seungjin-an replied:
Now that we removed the rewriting config logic, we no longer have "userStateIdToStoreIdMap". The change is that we have the nonUniqueStateId set from StateIdParser.scan(). Then, given this set, we'll create the stateIdToStoreMapping for each ParDo and initialize the SamzaStateStoreInternals, while keeping the stateId-> store map in "stores". Please take a look and let me know if this looks reasonable. Thank you!

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java:
```diff
@@ -105,6 +108,46 @@ public static void createConfig(
+   * ("foo", "First_Stateful_ParDo_with_same_stateId") storeId =
+   * "foo-First_Stateful_ParDo_with_same_stateId"
+   */
+  public static void rewriteConfigWithMultiParDoStateId(
```
Thank you :-)

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java:
```diff
@@ -83,6 +84,7 @@ public class TranslationContext {
   private final Map<PValue, String> idMap;
   private final Map<String, MessageStream> registeredInputStreams = new HashMap<>();
   private final Map<String, Table> registeredTables = new HashMap<>();
+  private final Set<String> multiParDoStateIds;
```
I really like "nonUniqueStateIds".. made the change throughout the diff. Thank you!

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java:
```diff
@@ -0,0 +1,29 @@
+   * Join stateId and escaped PTransform name, used for RocksDB storeId of stateIds with multiple
+   * ParDos.
+   */
+  public static String toMultiParDoStoreId(String stateId, String escapedPTransformName) {
```
Done.

xinyuiscool commented:
Thanks for fixing the problem for #1 in my previous comments. However, seems some of the necessary changes are also reverted for #2. I expect passing a mapping {stateId -> storeId} (currently it's a set of stateId) to the ```SamzaStoreStateInternals.createStateInternalsFactory()```. Then createStateInternalsFactory() can use that information to create the ```stores``` map.  Somehow this logic is missing. It was present in your previous code. Please take a look and run ./gradlew :runners:samza:build to verify the tests pass.

seungjin-an commented:
Thank you @xinyuiscool for the review again. SamzaStoreStateInternals logic has been included and verified the changes with all the tests passing from ./gradlew :runners:samza:build

xinyuiscool commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:
```diff
@@ -35,12 +34,13 @@ public class ConfigContext {
   private AppliedPTransform<?, ?, ?> currentTransform;
   private final SamzaPipelineOptions options;
-  private final Set<String> stateIds;
+  private final Set<String> nonUniqueStateIds;
```
Please take a look at the comments below to wrap this in a StoreIdGenerator class.

seungjin-an replied:
Done.

xinyuiscool commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java:
```diff
@@ -83,6 +83,7 @@ public class TranslationContext {
   private final Map<PValue, String> idMap;
   private final Map<String, MessageStream> registeredInputStreams = new HashMap<>();
   private final Map<String, Table> registeredTables = new HashMap<>();
+  private final Set<String> nonUniqueStateIds;
```
See comments above to wrap this in a StoreIdGenerator class.

xinyuiscool commented on runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java:
```diff
@@ -0,0 +1,26 @@
+ */
+package org.apache.beam.runners.samza.util;
+
+public class StoreIdUtils {
```
We can move this method in this class into the new StoreIdGenerator.

seungjin-an replied:
Done.

xinyuiscool reviewed (commented):
Overall looks great! Have some minor comments about consolidating some duplicate logic into a single class.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java:
```diff
@@ -377,17 +392,15 @@ public Map<String, String> createConfig(
-          throw new IllegalStateException(
-              "Duplicate StateId " + storeId + " found in multiple ParDo.");
+        final String stateId = state.id();
+        String storeId = stateId;
```
Awesome, changes are made.

seungjin-an commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java:
```diff
@@ -161,6 +163,17 @@ private static <InT, OutT> void doTranslate(
+    final DoFnSignature signature = DoFnSignatures.getSignature(transform.getFn().getClass());
+    final Map<String, String> stateIdToStoreMapping = new HashMap<>();
+    for (String stateId : signature.stateDeclarations().keySet()) {
+      String storeId = stateId;
```
Done.

seungjin-an commented:
> Overall looks great! Have some minor comments about consolidating some duplicate logic into a single class.

@xinyuiscool Awesome, I've addressed all the comments above. Please review upon your availability! Thank you.

xinyuiscool reviewed (approved):
LGTM!

xinyuiscool commented:
Run Java PreCommit
